pubsub/kafkapubsub/kafka.go

```diff
@@ -0,0 +1,553 @@
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewSubscription(ds, recvBatcherOpts, nil), nil
+}
+
+// openSubscription returns the driver for OpenSubscription. This function
+// exists so the test harness can get the driver interface implementation if it
+// needs to.
+func openSubscription(brokers []string, config *sarama.Config, group string, topics []string, opts *SubscriptionOptions) (driver.Subscription, error) {
+	if opts == nil {
+		opts = &SubscriptionOptions{}
+	}
+	consumerGroup, err := sarama.NewConsumerGroup(brokers, group, config)
+	if err != nil {
+		return nil, err
+	}
+	// Create a cancelable context for the background goroutine that
+	// consumes messages.
+	ctx, cancel := context.WithCancel(context.Background())
+	joinCh := make(chan struct{})
+	ds := &subscription{
+		opts:    *opts,
+		ch:      make(chan *sarama.ConsumerMessage),
+		closeCh: make(chan struct{}),
+		joinCh:  joinCh,
+		cancel:  cancel,
+	}
+	// Start a background consumer. It should run until ctx is cancelled
+	// by Close, or until there's a fatal error (e.g., topic doesn't exist).
+	// We're registering ds as our ConsumerGroupHandler, so sarama will
+	// call [Setup, ConsumeClaim (possibly more than once), Cleanup]
+	// repeatedly as the consumer group is rebalanced.
+	// See https://godoc.org/github.com/Shopify/sarama#ConsumerGroup.
+	go func() {
+		closeErr := consumerGroup.Consume(ctx, topics, ds)
+		ds.closeErr = closeErr
+		consumerGroup.Close()
+		close(ds.closeCh)
+	}()
+	if opts.WaitForJoin > 0 {
+		// Best effort wait for first consumer group session.
+		select {
+		case <-joinCh:
+		case <-ds.closeCh:
+		case <-time.After(opts.WaitForJoin):
+		}
+	}
+	return ds, nil
+}
+
+// Setup implements sarama.ConsumerGroupHandler.Setup. It is called whenever
+// a new session with the broker is starting.
+func (s *subscription) Setup(sess sarama.ConsumerGroupSession) error {
+	// The first time, close joinCh to (possibly) wake up OpenSubscription.
+	if s.joinCh != nil {
+		close(s.joinCh)
+		s.joinCh = nil
+	}
+	// Record the current session.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.sess = sess
+	return nil
+}
+
+// Cleanup implements sarama.ConsumerGroupHandler.Cleanup.
+func (s *subscription) Cleanup(sarama.ConsumerGroupSession) error {
+	// Clear the current session.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.sess = nil
+	return nil
+}
+
+// ConsumeClaim implements sarama.ConsumerGroupHandler.ConsumeClaim.
+// This is where messages are actually delivered, via a channel.
+func (s *subscription) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for msg := range claim.Messages() {
+		// We've received a message. Send it over the channel to ReceiveBatch.
+		// s.ch has no buffer, so the send blocks until ReceiveBatch is
+		// ready to receive them.
+		select {
+		case s.ch <- msg:
+		case <-sess.Context().Done():
+			// This session is over, we must return. We'll end up dropping msg, but
```

Comment:
I don't think so. Sarama will cancel this context when it wants the session to end. It probably expects that the `range` will end because it closed the channel of messages; in our case, that might happen, but we might also get to this case if we got a message from the `range` but are waiting to hand it off ot the user. It [looks like](https://github.com/Shopify/sarama/blob/master/consumer_group.go#L616) if we return an error it will try to do something with it.

It would be better if we could avoid getting the message at all until the user is going to read it; left a TODO.